tools/validate_spec.py

```diff
@@ -0,0 +1,314 @@
+    abort_on_error: bool,
+) -> None:
+    # pylint: disable=too-many-locals
+    def get_message_to_validate(path: Path, is_vaild: bool) -> "Message":
+        if not path.is_file():
+            raise TypeError(f"{path} is not a regular file")
+        message_bytes = path.read_bytes()
+        message = Message(message_bytes, is_vaild, path.name)
+        return message
+
+    if path_valid is None and path_invalid is None:
+        raise ValueError("no directory containing valid or invalid messages has been specified")
+
+    identifier = ID(root_message_id.split("::")).parts
+    if len(identifier) != 2:
+        raise NameError(f"supplied message identifier is not valid: {root_message_id}")
+    msg_name = identifier[0]
+    package_name = identifier[1]
+    try:
+        pdu_message = PyRFLX.from_specs([str(path_spec)])[msg_name][package_name]
+    except KeyError:
+        raise KeyError(f"message {msg_name} could not be found in package {package_name}")
+
+    path_generators: Dict[Path, Iterator[Path]] = {}
+    for path in [path_valid, path_invalid]:
+        if path is not None:
+            path_generators[path] = path.glob("*")
+
+    json_output_writer: Optional[JsonOutputWriter] = None
+    if full_output is not None:
+        json_output_writer = JsonOutputWriter(full_output)
+
+    validator = Validator(pdu_message)
+    for k, path_generator in path_generators.items():
+        for message_file_path in path_generator:
+            msg = get_message_to_validate(message_file_path, k == path_valid)
+            parser_result = validator.parse_message(msg.bytes)
+            validation_result = validator.validate_message(msg, parser_result)
+            if json_output_writer is not None:
+                json_output_writer.write(validation_result.get_json_output())
+            print(validation_result.get_abbreviated_output())
+            if abort_on_error and validation_result.get_classification() in [
+                Classification.FP,
+                Classification.FN,
+            ]:
+                if json_output_writer is not None:
+                    json_output_writer.close()
+                sys.exit(1)
+
+    if json_output_writer is not None:
+        json_output_writer.close()
+
+
+class JsonOutputWriter:
+    def __init__(self, file: Path) -> None:
+        self.file = open(file, "w+")
+        self.file.write("[\n")
+        self.count = 0
+
+    def write(self, content: Dict[str, object]) -> None:
+        if self.count != 0:
+            self.file.write(",\n")
+        json.dump(
+            content,
+            self.file,
+            indent="\t",
+        )
+        self.count += 1
+
+    def close(self) -> None:
+        self.file.write("\n]")
+        self.file.close()
+
+
+class Validator:
+    def __init__(self, pdu_message: MessageValue) -> None:
+        self.pdu_message = pdu_message
+
+    def get_pdu_message(self) -> MessageValue:
+        return self.pdu_message.clone()
```

Comment:
Why is that a public method? Is there a need for this method at all?

Reply:
Removed in 7a4baa8